File: Lenguajes de programacion/cppcalc/parser.cpp
#include "parser.h"
#include "calcex.h"
#include <string>
#include <sstream>
using namespace std;

Parser::Parser(istream* in) {
  scan = new Scanner(in);
}

Parser::~Parser() {
  try {
    delete scan;
  } catch (...) {}
}

AST* Parser::parse() {
  return Prog();
}

AST* Parser::Prog() {
  AST* result = Expr();
  Token* t = scan->getToken();

  if (t->getType() != eof) {
    cout << "Syntax Error: Expected EOF, found token at column " << t->getCol() << endl;
    throw ParseError;
  }

  return result;
}

AST* Parser::Expr() {
  return RestExpr(Term());
}

AST* Parser::RestExpr(AST* e) {
  Token* t = scan->getToken();

  if (t->getType() == add) {
    return RestExpr(new AddNode(e,Term()));
  }

  if (t->getType() == sub)
    return RestExpr(new SubNode(e,Term()));

  scan->putBackToken();

  return e;
}

AST* Parser::Term() {
  return RestTerm(Storable());
  }

AST* Parser::RestTerm(AST* e) {
  Token* t = scan->getToken();

  if (t->getType() == times) {
    return RestTerm(new TimesNode(e,Storable()));
  }

  if (t->getType() == divide){
    return RestTerm(new DivideNode(e,Storable()));
  }
  //Agregado el RestTerm para modulo
  if (t->getType() == mod){
    return RestTerm(new ModNode(e,Storable())); 
  }
   
  scan->putBackToken();
  return e;

}
//Storable modificado para MemOperation
AST* Parser::Storable() {

  AST* result = MemOperation(Factor());
  return result;
}

//Agreando MemOperation para mas operaciones de S, P y M
AST* Parser::MemOperation(AST* result){

  Token* t = scan->getToken();

  if(t->getType() == keyword){
    if(t->getLex() == "S") 
      return new StoreNode(result);
    //Caso para PlusNode
    else if(t->getLex() == "P")
      return new PlusNode(result);
    //Caso para MinusNode
    else if(t->getLex() == "M")
      return new MinusNode(result);
  }
  scan->putBackToken();

  return result;
}

  
AST* Parser::Factor() {
  Token* t = scan->getToken();

  if (t->getType() == number) {
    istringstream in(t->getLex());
    int val;
    in >> val;
    return new NumNode(val);
  }
  //Se agrego el tipo identifier
  if(t->getType()==identifier){
    scan->putBackToken();
    return Assignable();
  }

  if(t->getType() == keyword) {
    if(t->getLex() == "R"){
      return new RecallNode();
    }
    //Para el caso de clear
    else if(t->getLex() == "C"){
      return new ClearNode();
    }
    else{
      cout << "Syntax error keyword: "
	   << t->getLex() << " at line: " << t->getLine()

	   << " col: " << t->getCol() << endl;
      throw ParseError;
    }
  }

     
  if(t->getType() == lparen){
    AST *result = Expr();
    t = scan->getToken();
    if(t->getType() == rparen){
      return result;
    }
    else{
      //Aqui va un error
      cout<<"sintax error expected ')' found"
	  << t ->getType() << " at line: " << t->getLine()
	  <<" col: " << t->getCol() << endl;
      throw ParseError;
    }
  }    
     
  cout << "Expected number, R or '('" << endl;
  throw ParseError;
}

//Agregado para las asignaciones
AST* Parser::Assignable(){
  Token* t = scan->getToken();
  if((t->getType())==identifier){
    AST* result = Assign(t->getLex());
    return result;
  }else{
    //cout identifier expcted
  }
}

AST* Parser::Assign(string s){
  
  Token* t= scan->getToken();
  if(t->getType() == equ){
    return new AssignNode(s,Expr());
  }
  else{
    scan->putBackToken();
    return new IdNode(s);
  }
}

File: Lenguajes de programacion/rubycalc/lib/parser.rb
# coding: utf-8
require 'ast'
require 'scanner'
require 'token'
require 'calcex'

class Parser
  def initialize(istream)
    @scan = Scanner.new(istream)
  end
   
  def parse()
    return Prog()
  end
  
  private
  def Prog()
    result = Expr()
    #t = @scan.getToken
    
    #if t.type != :eof then
    #  print "Expected EOF. Found ", t.type, ".\n"
    #  raise ParseError.new
    #end
    
   return result
  end
  
  def Expr() 
    RestExpr(Term())
  end
   
  def RestExpr(e) 
    t = @scan.getToken
    
    if t.type == :add then
      return RestExpr(AddNode.new(e,Term()))
    end
    
    if t.type == :sub then
      return RestExpr(SubNode.new(e,Term()))
    end
      
    @scan.putBackToken()
    return e 
  end
  
  def Term()
    RestTerm(Storable())
  end

    
  def RestTerm(e)

    t = @scan.getToken
    if t.type == :times then
      return RestTerm(TimesNode.new(e,Storable()))
    end
    
    if t.type == :divide then
      return RestTerm(DivideNode.new(e,Storable()))
    end
    #Agregado caso modulo
    if t.type == :mod then
      return RestTerm(ModNode.new(e,Storable()))
    end
    
    @scan.putBackToken()
    return e
  end
   
  def Storable()
    MemOperation(Factor())    
  end

  def MemOperation(result)
    t = @scan.getToken

    if t.type == :keyword then
      if t.lex == 'S' then
        return StoreNode.new(result)
      end
      if t.lex == 'P' then
        return PlusNode.new(result)
      end
      if t.lex == 'M' then
        return MinusNode.new(result)
      end
      puts "Expected S, found: #{t.lex} at line #{t.line}, col: #{t.col}"
      raise ParseError.new
    end

    @scan.putBackToken()

    return result
  end
    
  def Factor() 

    t = @scan.getToken
    if t.type == :number then
      return NumNode.new(t.lex.to_i)
    end

    if t.type == :identifier then
      @scan.putBackToken
      return Assignable()
    end
    
    if t.type == :keyword then
      if t.lex == 'R' then
        return RecallNode.new
      elsif t.lex == 'C' then
        return ClearNode.new
      end
      puts "Expected R, found #{t.lex} at line: #{t.line}, col: #{t.col}"
      r
      aise ParseError.new
    end

    if t.type == :lparen then
      result = Expr()
      t = @scan.getToken
      if t.type == :rparen then
        return result
      end

      puts "Parser Error: expected ) found: #{t.lex}, at line: #{t.line}, col: #{t.col}"
      raise ParseError.new
    end

    puts "Parser Error: expected number, R or ( found: #{t.lex}, at line: #{t.line}, col: #{t.col}"
    raise ParseError.new
  end
  
  def Assignable()
    t=@scan.getToken
    return Assign(t.lex)
  end

  def Assign(str)
    t=@scan.getToken
    if t.type == :equ
      return AssignNode.new(str, Expr())
    end
    @scan.putBackToken
    return IdNode.new(str)
  end
end
